substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,526 @@
+	///
+	/// IMPORTANT: **Checks made in validation need not be repeated here.**
+	fn prepare(
+		self,
+		val: Self::Val,
+		origin: &OriginOf<Call>,
+		call: &Call,
+		info: &DispatchInfoOf<Call>,
+		len: usize,
+		context: &Context,
+	) -> Result<Self::Pre, TransactionValidityError>;
+
+	/// Do any post-flight stuff for an extrinsic.
+	///
+	/// `_pre` contains the output of `prepare`.
+	///
+	/// This gets given the `DispatchResult` `_result` from the extrinsic and can, if desired,
+	/// introduce a `TransactionValidityError`, causing the block to become invalid for including
+	/// it.
+	///
+	/// Parameters:
+	/// - `pre`: `Self::Pre` returned by the result of the `prepare` call prior to dispatch.
+	/// - `info`: Information concerning, and inherent to, the transaction's call.
+	/// - `post_info`: Information concerning the dispatch of the transaction's call.
+	/// - `len`: The total length of the encoded transaction.
+	/// - `result`: The result of the dispatch.
+	/// - `context`: Some opaque mutable context, as yet unused.
+	///
+	/// WARNING: It is dangerous to return an error here. To do so will fundamentally invalidate the
+	/// transaction and any block that it is included in, causing the block author to not be
+	/// compensated for their work in validating the transaction or producing the block so far. It
+	/// can only be used safely when you *know* that the transaction is one that would only be
+	/// introduced by the current block author.
+	fn post_dispatch(
+		_pre: Self::Pre,
+		_info: &DispatchInfoOf<Call>,
+		_post_info: &PostDispatchInfoOf<Call>,
+		_len: usize,
+		_result: &DispatchResult,
+		_context: &Context,
+	) -> Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	/// Compatibility function for supporting the `SignedExtension::validate_unsigned` function.
+	///
+	/// DO NOT USE! THIS MAY BE REMOVED AT ANY TIME!
+	#[deprecated = "Only for compatibility. DO NOT USE."]
+	fn validate_bare_compat(
+		_call: &Call,
+		_info: &DispatchInfoOf<Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		Ok(ValidTransaction::default())
+	}
+
+	/// Compatibility function for supporting the `SignedExtension::pre_dispatch_unsigned` function.
+	///
+	/// DO NOT USE! THIS MAY BE REMOVED AT ANY TIME!
+	#[deprecated = "Only for compatibility. DO NOT USE."]
+	fn pre_dispatch_bare_compat(
+		_call: &Call,
+		_info: &DispatchInfoOf<Call>,
+		_len: usize,
+	) -> Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	/// Compatibility function for supporting the `SignedExtension::post_dispatch` function where
+	/// `pre` is `None`.
+	///
+	/// DO NOT USE! THIS MAY BE REMOVED AT ANY TIME!
+	#[deprecated = "Only for compatibility. DO NOT USE."]
+	fn post_dispatch_bare_compat(
+		_info: &DispatchInfoOf<Call>,
+		_post_info: &PostDispatchInfoOf<Call>,
+		_len: usize,
+		_result: &DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		Ok(())
+	}
+}
+
+/// Implict
+#[macro_export]
+macro_rules! impl_tx_ext_default {
```

Comment:
Not really sure we should have this macro.

But if we want to have it, it should have proper docs.

Reply:
It's just a convenience macro to avoid writing boilerplate. It should state what the default impl is, I agree.

Reply:
I mean I see what it does. However, users will not know what it does and what the expected syntax is.